### ISSUE TITLE
Problem with reloadOnSearch: false - change in stateParams doesn't reload 

### ISSUE THREAD

**Issue by OriaLogic** [angular-ui/ui-router#1079]
Here is my issue:

I navigate through a list of links, that display different images depending on the date in the search parameters (?from&to). 

Those search parameters should not trigger a state reload. Finally, here is my state declaration:
```javascript
state('links/:linkId', {
  url: "/links/linkId?from&to",
  reloadOnSearch: false,
  templateUrl: ...
  Controller: ...
});
```

But then a huge problem shows up! When I try to navigate through my different links, reload doesn't happen anymore, so that I cannot pass from one link to another.

After some digging in the code, I found out that when `reloadOnSearch` equals to false, triggers are always blocked because there's absolutely no checking of whether the changing parameters are search parameters or path parameters.

So actually, `reloadOnSearch` doesn't work as intended and should be renamed `reloadOnStateParamsChange`.

I fixed the issue by checking if the state parameters that changed are search parameters or not, which seems to work fine.

Please tell me if I'm mistaking somewhere or if it is relevant, and also if you would be interested in a pull request or a plunkr. 

Thanks in advance!

**Comment by nawlbergs** [comment]
I can also confirm this issue. I'm using search parameters to store minor ui variables within a route/view.  
e.g.
#/user/123/reports?from=today&to=tomorrow
and cannot change the users id.

**Comment by UmairB** [comment]
My hilarious hack for this is as follows

//set the reloadOnSearch on state to false initially in some code block
$state.current.reloadOnSearch = false;

// update query string here
$location.search({"from": 1, "to": 2});

//revert reloadOnSearch on the state!
$timeout(function () {
    $state.current.reloadOnSearch = undefined;
});

**Comment by jpmckearin** [comment]
@nateabele @christopherthielen I noticed dynamic parameters has yet to be merged into master. Is there any reason that I can't be of assistance and just do the merge and submit a PR?

**Comment by christopherthielen** [comment]
@jpmckearin yes, unfortunately we're gated on the 1.0 refactor before we add any additional major features

**Comment by straker** [comment]
So I know there is a planned "fix" with dynamic parameters, but if you're gated for an undetermined amount of time from merging it in, could we at least get the fix of having reloadOnSearch actually reload on path changes now?

**Comment by mjrk** [comment]
+1

**Comment by eddiemonge** [comment]
yeah a patch fix would be nice. i know that makes refactors hard but it is a long standing bug

**Comment by lougaou** [comment]
+1

**Comment by skeisman** [comment]
This is a very important feature for my project.

If I make several changes to search params and then navigate to a child state, when I call $state.go('^') only the first search params are remembered.

I have spend several hours reading through the source code to try to fix this myself but it seems like it would take me a while to get up to speed. 

BTW thanks for creating such a useful router.

**Comment by skeisman** [comment]
When I use ui-sref on a page that has reloadOnSearch:false to change one of the search params clicking the <a> tag does nothing. This seems like a related bug.

**Comment by eddiemonge** [comment]
Why do you have a normal href link and not a ui-sref link?

**Comment by skeisman** [comment]
The bug I have described above also fails with ui-sref. I wanted to describe the problem with as few moving parts as possible. UI router ignores href clicks that do not change the state when you have reloadOnSearch:false

**Comment by eddiemonge** [comment]
Right, thats what reloadOnSearch: false means, don't reload the state when a query param changes.

**Comment by skeisman** [comment]
When I click an href link the url does not change when reloadOnSearch:false. That means reloadOnSearch:false is breaking normal browser functionality.

My specific problem aside A better way to make reloadOnSearch:false work is that the state should change but the view would not reload. This will cause $stateParams to be correct. This will cause $state.go('^') to work correctly and a whole bunch of other edge cases to work correctly. In its current state reloadOnSearch:false is nearly unusable because it breaks so many other things.

**Comment by ceoaliongroo** [comment]
I my case, i works for me create a decorator to keep sync ``$urlRouter`` and ``$location`` 

```javascript
    /**
       * Update update params values with reloadOnSearch:false.
       *
       * @param params
       *  Object of the change for the $stateParams.
       */
      $delegate.refreshUrlWith = function refreshUrlWith(params) {
        $location.search(extend(copy($stateParams),
          $location.search(),
          params
        ));
        $urlRouter.update(true);
      };
```

**Comment by metamatt** [comment]
I just sent a PR for this issue which I believe is a correct fix for the 0.2 codebase as it stands. The dynamic parameter stuff, from what I've been able to glean from in-passing references in this and a few other issue reports and PRs, looks pretty cool and I look forward to being able to use that, but I hope we can take this fix for the interim period before that lands.

**Comment by christopherthielen** [comment]
@metamatt  Thanks for the PR.  I modified your commit and merged it in manually.

**Comment by christopherthielen** [comment]
This plunkr demonstrates 0.2.15 reloadOnSearch functioning:  http://plnkr.co/edit/5OPoKUDnn98JiKFlyEYL?p=preview

**Comment by lambertsteven** [comment]
Works for me, awesome job @metamatt!

**Comment by drothlis** [comment]
> What will replace reloadOnSearch is dynamic parameters, which is currently partially implemented in master, and will be fully-realized in the next release, this will give you per-parameter (query string parameters or normal URL parameters) disabling of controller reload, and the ability to watch parameter changes.

@nateabele where is this documented? I don't see any mention of it at https://angular-ui.github.io/ui-router/site/#/api/ui.router.util.type:Type

**Comment by nateabele** [comment]
@drothlis Not super well-documented, sadly: https://christopherthielen.github.io/ui-router/1.0-alpha01/classes/params.param.html#dynamic

You define one like this:

```javascript
$stateProvider.state("foo", {
  url: "/:bar",
  params: {
    bar: { dynamic: true }
  }
});
```

**Comment by christopherthielen** [comment]
this is the temporary home for 1.0 alpha docs: http://angular-ui.github.io/ui-router/feature-1.0 

http://angular-ui.github.io/ui-router/feature-1.0/interfaces/params.paramdeclaration.html

**Comment by drothlis** [comment]
@nateabele, @christopherthielen: Thanks, that's very helpful! :-)

**Comment by LipillaiDave** [comment]
Is this in 1.5.9? 
Reloading the template is not acceptable to me because I've got a (hopefully) persistent STOMP controller that is in constant communication. When I switch from /item/2 to /item/3 I'd like the user to be able to bookmark that. Instead the STOMP controller (which is its own controller, so doesn't get reloaded) goes through all the steps of initiating communications again because the template gets reloaded.

My app.js includes this:
	.state('thing, {
		url: '/thing/:id',
		templateUrl: '/tpl/thing.html',
		params: { id: { dynamic: true }}
	});
Did I get something wrong? I'm running 1.5.8 at the moment.

Thanks.

**Comment by christopherthielen** [comment]
That looks fine. It shouldn't reload when the parameter changes. Are you using ui-router 1.0 beta? Dynamic params are not in the legacy versions.

**Comment by LipillaiDave** [comment]
@christopherthielen Thank you. I use Bower for most of my packages, and it seems like for some reason they only have the legacy version on Bower. NPM'd ui-router 1.0 beta works like a charm!
